asyn/drvAsynSerial/drvAsynIPPort.c

```diff
@@ -505,13 +505,56 @@ connectIt(void *drvPvt, asynUser *pasynUser)
             }
         }
 
+    }
+
+#ifdef USE_POLL
+    if (setNonBlock(fd, 1) < 0) {
+        epicsSnprintf(pasynUser->errorMessage,pasynUser->errorMessageSize,
+                               "Can't set %s O_NONBLOCK option: %s",
+                                       tty->IPDeviceName, strerror(SOCKERRNO));
+        epicsSocketDestroy(fd);
+        return asynError;
+    }
+#endif
+
+    if (pasynUser->reason <= 0) {
+
         /*
          * Connect to the remote host
          * If the connect fails, arrange for another DNS lookup in case the
          * problem is just that the device has DHCP'd itself an new number.
          */
         if (tty->socketType != SOCK_DGRAM) {
-            if (connect(fd, &tty->farAddr.oa.sa, (int)tty->farAddrSize) < 0) {
+            int connectResult = connect(fd, &tty->farAddr.oa.sa, (int)tty->farAddrSize);
+        #ifdef USE_POLL
+            if (connectResult < 0 && ((SOCKERRNO == EWOULDBLOCK) || (SOCKERRNO == EINPROGRESS))) {
+                double connectTimeout;
+                int msConnectTimeout;
+                struct pollfd pollfd;
+
+                pasynManager->getAutoConnectTimeout(&connectTimeout);
+                msConnectTimeout = 1000 * connectTimeout;
```

Comment:
One thing to note, which I don't know if it's relevant, is that any timeouts below 1ms will be rounded down to 0, causing poll to return immediately.

I don't think anyone actually would have such strict timeout requirements, though.